fix: remove links for each level of the namespaces hierarchy pages in global graph view

8N9KT opened the pull request:
in reference to [#2964](https://github.com/logseq/logseq/issues/2964)

**graph example**
actual solution:
![patch_1657548238582_0](https://user-images.githubusercontent.com/84566861/180021556-fb45cd3f-4bc6-4995-8ac8-dff973fb6489.png)

with patch:
![patch_1657548296655_0](https://user-images.githubusercontent.com/84566861/180021715-20c5332e-a08d-4c2f-ae4a-4948d2c8a437.png)

**Question:**
is it necessary to do the same also in the page graph?

cnrpman commented:
Thank you! Agree that it's a disaster that having these namespaces as ref-pages.
But I think it should be resolved in the db structure instead of filtering the edges..

8N9KT commented:
@cnrpman 
> Thank you! Great work! Agree that it's a disaster that having these namespaces as ref-pages. But I'm concerning about the complexity of the filtering logic. Instead of filtering them out, I prefer not to include these > 2 hops of namespaces into `:block/ref` at start..

ok, how can i do it? ;)

cnrpman commented:
> @cnrpman
> 
> > Thank you! Great work! Agree that it's a disaster that having these namespaces as ref-pages. But I'm concerning about the complexity of the filtering logic. Instead of filtering them out, I prefer not to include these > 2 hops of namespaces into `:block/ref` at start..
> 
> ok, how can i do it? ;)

I made a naive try by tuning the query in `get-pages-relation`, utilizing the [`namespace` query rule](https://github.com/cnrpman/logseq/blob/2db3be5f313d7fcd423fc63b699713ff8c6171d8/deps/db/src/logseq/db/rules.cljc#L20-L24) to do recursive check:
```clojure
(defn get-pages-relation
  [repo with-journal?]
  (when-let [db (conn/get-db repo)]
    (let [q (if with-journal?
              '[:find ?page ?ref-page-name
                :in $ %
                :where
                [?p :block/name ?page]
                [?block :block/page ?p]
                [?block :block/refs ?ref-page]
                ;; allow direct namespace edges, disallow indirect edges
                (or [?p :block/namespace ?ref-page]
                    (not (namespace ?ref-page ?p)))
                [?ref-page :block/name ?ref-page-name]]
              '[:find ?page ?ref-page-name
                :in $ %
                :where
                [?p :block/journal? false]
                [?p :block/name ?page]
                [?block :block/page ?p]
                [?block :block/refs ?ref-page]
                ;; allow direct namespace edges, disallow indirect edges
                (or [?p :block/namespace ?ref-page]
                    (not (namespace ?ref-page ?p)))
                [?ref-page :block/name ?ref-page-name]])]
      (->>
       (d/q q db rules)
       (map (fn [[page ref-page-name]]
              [page ref-page-name]))))))
```
<img width="786" alt="image" src="https://user-images.githubusercontent.com/9862022/180369533-4a8f62b0-bbb3-4bd2-aeab-009044d85489.png">

cnrpman commented:
@8N9KT I made some attempts in `get-pages-relation` with `not-join` but failed.. So the filter solution is good to me 🤣

logseq-cldwalker commented on src/main/frontend/db/model.cljs:
```diff
@@ -1094,6 +1094,47 @@
      (distinct))))
 
+; return list of parents namespace
+(defn get-parents-namespace-list [page-namespace & nested-found]
```
All these added fns should be marked as private `defn-` since they aren't used outside this ns

8N9KT replied:
> I made some attempts in the `get-pages-relation` query with `not-join` but failed.. So the filter solution is good to me rofl

> Thanks for improving this. I'm also concerned about the complexity of maintaining this.

hello @cnrpman  and @logseq-cldwalker, thank you for giving me your time.
we all know that the solution I have implemented is not the best but with the `:block/ref` that are now available filtering them is the only solution.

if you agree I open a "discussions" to propose my idea on how to solve this problem.

meanwhile I continue with this solution...

logseq-cldwalker commented on src/main/frontend/db/model.cljs:
```diff
@@ -1094,6 +1094,47 @@
       (util/safe-page-name-sanity-lc page))
      (distinct))))
 
+; return list of parents namespace
```
Great to see docs for your fns. Would be good to associate this with a fn. See `get-paginated-blocks-no-cache` in this file for an example

logseq-cldwalker commented on src/main/frontend/db/model.cljs:
```diff
@@ -1094,6 +1094,47 @@
 
+; return list of parents namespace
+(defn get-parents-namespace-list [page-namespace & nested-found]
+  (if (string/includes? page-namespace "/") ;isNamespace
```
We should move any references to "/" outside fns to a separate `(def namespace-char "/")`. I'm surprised we don't already have this var defined elsewhere. You can create a regex dynamically with `(re-pattern namespace-char)`

8N9KT replied:
> I made some attempts in the `get-pages-relation` query with `not-join` but failed.. So the filter solution is good to me rofl

> Thanks for improving this. I'm also concerned about the complexity of maintaining this.

hello @cnrpman  and @logseq-cldwalker, thank you for giving me your time.
we all know that the solution I have implemented is **not the best** but with the `:block/ref` that are now available filtering them is the **only** solution.

if you agree I open a "discussions" to propose my idea on how to solve this problem.

meanwhile I continue with this solution...

cnrpman replied:
Can you use this?
https://github.com/logseq/logseq/blob/70acbadc70ef567ad96c63513a406a7118588d45/deps/graph-parser/src/logseq/graph_parser/text.cljs#L217

logseq-cldwalker reviewed (changes requested):
@8N9KT Thanks for improving this. I'm also concerned about the complexity of maintaining this. Like #6120, it would be helpful to have some unit tests that demonstrate that this works e.g. only the specific namespace child/leaf gets associated. Testing `get-pages-relation` is preferred and may be easier to setup thanks to the `load-test-files` test helper. But you could also write one for `remove-nested-namespaces-link` if you'd prefer

> is it necessary to do the same also in the page graph?

I wouldn't worry about this yet. One PR at a time :)

8N9KT commented:
> I made some attempts in the `get-pages-relation` query with `not-join` but failed.. So the filter solution is good to me rofl

> Thanks for improving this. I'm also concerned about the complexity of maintaining this.

hello @cnrpman  and @logseq-cldwalker, thank you for giving me your time.
we all know that the solution I have implemented is **not the best** but with the `:block/ref` that are now available filtering them is the **only** solution.

if you agree I open a "discussions" to propose my idea on how to solve this problem.

meanwhile I continue with this solution...

8N9KT commented:
> Testing `get-pages-relation` is preferred and may be easier to setup thanks to the `load-test-files` test helper. But you could also write one for `remove-nested-namespaces-link` if you'd prefer

hi @logseq-cldwalker, in which file i can add this unit tests?

logseq-cldwalker commented:
> hi @logseq-cldwalker, in which file i can add this unit tests?

Hi. Unit tests would go in model_test.cljs. I've updated docs in https://github.com/logseq/logseq/blob/master/docs/dev-practices.md#unit-testing for explaining the general rule of thumb

logseq-cldwalker commented:
> if you agree I open a "discussions" to propose my idea on how to solve this problem.

Opening a discussion sounds great

8N9KT commented:
> > hi @logseq-cldwalker, in which file i can add this unit tests?
> 
> Hi. Unit tests would go in model_test.cljs. I've updated docs in https://github.com/logseq/logseq/blob/master/docs/dev-practices.md#unit-testing for explaining the general rule of thumb



> > hi @logseq-cldwalker, in which file i can add this unit tests?
> 
> Hi. Unit tests would go in model_test.cljs. I've updated docs in https://github.com/logseq/logseq/blob/master/docs/dev-practices.md#unit-testing for explaining the general rule of thumb

hi @logseq-cldwalker  how can i test the private function?

logseq-cldwalker commented:
Hi. You can test them by calling them as (#'namespace/var e.g.
(#'model/get-parents-namespace-list ...)

On Thu, 28 Jul 2022 at 09:41, 8N9KT ***@***.***> wrote:

> hi @logseq-cldwalker <https://github.com/logseq-cldwalker>, in which file
> i can add this unit tests?
>
> Hi. Unit tests would go in model_test.cljs. I've updated docs in
> https://github.com/logseq/logseq/blob/master/docs/dev-practices.md#unit-testing
> for explaining the general rule of thumb
>
> hi @logseq-cldwalker <https://github.com/logseq-cldwalker>, in which file
> i can add this unit tests?
>
> Hi. Unit tests would go in model_test.cljs. I've updated docs in
> https://github.com/logseq/logseq/blob/master/docs/dev-practices.md#unit-testing
> for explaining the general rule of thumb
>
> hi @logseq-cldwalker <https://github.com/logseq-cldwalker> how can i test
> the private function?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/logseq/logseq/pull/6115#issuecomment-1198159003>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXFVC56X6JKAKZ4WZ3HBNI3VWKEYBANCNFSM54EDRAXQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cnrpman reviewed (approved):
Thank you for adding test cases!
LGTM

logseq-cldwalker reviewed (approved):
@8N9KT Thanks for fixing with tests! 👍 🚢 I had some minor cleanup but I'll just push them directly. In the future, if you select [Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork), we can push to your PR

CodingXCat commented:
> in reference to [#2964](https://github.com/logseq/logseq/issues/2964)
> 
> **graph example** actual solution: ![patch_1657548238582_0](https://user-images.githubusercontent.com/84566861/180021556-fb45cd3f-4bc6-4995-8ac8-dff973fb6489.png)
> 
> with patch: ![patch_1657548296655_0](https://user-images.githubusercontent.com/84566861/180021715-20c5332e-a08d-4c2f-ae4a-4948d2c8a437.png)
> 
> **Question:** is it necessary to do the same also in the page graph?

I have updated to the latest version and excited to see this feature which have troubled me for a long time.
I think page graph is also need this feature, because namespace is the only way to do directory-like organize in logseq now, and page graph is more useful when do link walking.
And more if logseq could expand links which link-depth > 2 in page graph(which could be changed in the panel), then links to the root could be visualized by this way.
Combined, this will be a good way to do directory-like or tag-category organize in logseq.

Thanks again for your great feature~
